core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -431,10 +428,9 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
   def updateWorkspace(workspaceName: WorkspaceName, operations: Seq[AttributeUpdateOperation]): Future[PerRequestMessage] = {
     withAttributeNamespaceCheck(operations.map(_.name)) {
       for {
+        ctx <- getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write)
         workspace <- dataSource.inTransaction({ dataAccess =>
-          withWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write, dataAccess) {
-            updateWorkspace(operations, dataAccess)
-          }
+            updateWorkspace(operations, dataAccess)(ctx)
```

Comment:
by changing the transaction boundaries, this opens up the small chance of a race condition. We get the workspace in transaction A, then update the workspace in transaction B. Between those transactions, some other user could sneak in and perform a different update, which could get overwritten in transaction  B. Is this an acceptable tradeoff? Is it worth protecting against?

Reply:
(Drive-by comment because I like thinking about this type of thing)

In a lot of cases, last-in-wins can be fine. I can think of two cases where it's not:

1. If other user's decision to modify is based on information that the first user is modifying which invalidates other user's decision
2. If only the changed fields are updated in such a way that two users making separate updates could leave the entity in an inconsistent/invalid state

That said, if we're not doing optimistic locking with a rev column or something like that (I don't remember if we are) then we're already vulnerable to these problems. So the answer to "is it worth protecting against" depends on the fields that can be modified and whether this specific type of entity is vulnerable to either of these two scenarios. Can you think of any other potentially dangerous scenarios?

Reply:
note the last commit - 7f5b533e841ca70b7c6befe9d32b7d69749e7069 - in which I've re-retrieved the workspace for `updateLibraryAttributes`, `updateWorkspace`, and `cloneWorkspace` to avoid the race condition.

For these three APIs, this trades an extra db query to re-get the workspace for removing the Sam call from a transaction. I think this is worth it ... do you?

Reply:
Yes